lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -1255,9 +1252,16 @@ matchWitness(TypeChecker &tc,
   };
 
   // Match a type in the requirement to a type in the witness.
-  auto matchTypes = [&](Type reqType, Type witnessType) 
+  auto matchTypes = [&](Type reqType, Type witnessType)
                       -> Optional<RequirementMatch> {
-    cs->addConstraint(ConstraintKind::Equal, reqType, witnessType, locator);
+    ConstraintKind kind = ConstraintKind::Conversion;
+    if (auto func = dyn_cast<FuncDecl>(req)) {
+      auto name = req->getFullName().getBaseName();
+      if (func->isOperator() && name.str() == "==") {
+        kind = ConstraintKind::Equal;
+      }
+    }
+    cs->addConstraint(kind, witnessType, reqType, locator);
```

Comment:
Because relaxing `ConstraintKind` isn't enough.

In this example:
```swift
protocol P {
    var a: Int? { get }
}

struct S: P {
    var a: Int
}
```
Constraint system would try to match `Optional<Int>` (as the requirement) to `Int` (as witness) and it would fail.

In contrast to function type convertion:
```swift
func f() -> Int { return 3 }
let g : ()->Int? = f
```
Constraint system would try to match `Int` to `Optional<Int>`, which works fine.